### PR TITLE
[Merged by Bors] - Support marriages and merged ATXs in checkpoint-recovery

### PR DESCRIPTION
**Pull request by poszu**
## Motivation

Persist marriages (which IDs are married and form an equivocation set) and marriage ATXs (so they can be referenced in future ATXs) across the checkpoint-restore.

## Test Plan

Added tests.

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: checkpoint/collecting_deps_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/spacemeshos/go-spacemesh/activation/wire"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/fixture"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+)
+
+func TestCollectingDeps(t *testing.T) {
```
Comment (fasmat) [inline checkpoint/collecting_deps_test.go]: I'm not sure about the filename: could we change it to `recovery_collecting_deps_test.go`?

File: activation/e2e/checkpoint_merged_test.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+	mFetch.EXPECT().RegisterPeerHashes(peer.ID(""), gomock.Any())
+	mFetch.EXPECT().GetPoetProof(gomock.Any(), gomock.Any())
+	mFetch.EXPECT().GetAtxs(gomock.Any(), gomock.Any(), gomock.Any())
+	mBeacon.EXPECT().OnAtx(gomock.Any())
+	mTortoise.EXPECT().OnAtx(gomock.Any(), gomock.Any(), gomock.Any())
+	err = atxHdlr.HandleGossipAtx(context.Background(), "", codec.MustEncode(mergedATX))
+	require.NoError(t, err)
+
+	// Step 3. Checkpoint
```
Comment (fasmat) [inline activation/e2e/checkpoint_merged_test.go:241]: Everything before this I think we don't really need to test e2e - we could just start here with a DB containing all the data that was generated up until this point. The point of this test is to verify that checkpoint is working e2e not the atxbuilder or handler. wdyt?

File: activation/e2e/checkpoint_merged_test.go [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+	// 3.1. Generate checkpoint
+	require.NoError(t, accounts.Update(db, &types.Account{}))
```
Comment (fasmat) [inline activation/e2e/checkpoint_merged_test.go:243]: Why is this needed?

Comment (poszu) [reply activation/e2e/checkpoint_merged_test.go:243]: Because checkpoint generation will fail if there are no accounts. IDK if this is a bug or a feature but there is even a unit test covering this scenario and asserting that `Generate()` fails: https://github.com/spacemeshos/go-spacemesh/blob/679a98d74da9c6ee4effa50786d6def9c57ff660/checkpoint/runner_test.go#L362-L374

File: activation/e2e/checkpoint_merged_test.go (hunk 2 above)
Comment (poszu) [reply activation/e2e/checkpoint_merged_test.go:241]: That would work too, but on the other hand, the test would need to assume much about the implementation (it would need to put atxs and marriages into the DB). A fully e2e test using the exposed API surface (ATX handler, ATX wire types) has less chance of breaking in the future when the internal implementation changes.

**Review by fasmat — APPROVED** [review]

**Comment by poszu** [comment]
bors merge
